scripts/acceptance-test/src/state_consistency.rs

```diff
@@ -0,0 +1,296 @@
+                tracing::warn!(
+                    "State assertion tx for slot {visible_slot} height {rollup_height} was rejected because the rollup already advanced past it. This can happen, but the kernel assertions have been skipped for this slot as a result."
+                );
+                return Ok(());
+            }
+
+            anyhow::bail!(
+                "Failed to submit state assertion tx for slot {visible_slot} height {rollup_height}: {e}",
+            );
+        }
+    }
+}
+
+/// Check if the error is a "too late" error, meaning the transaction was rejected
+/// because the actual rollup height is greater than the expected height.
+fn is_too_late_error(
+    e: &sov_api_spec::Error<sov_api_spec::types::ApiError>,
+    expected_height: u64,
+) -> bool {
+    // Extract the ApiError from the progenitor Error wrapper
+    let api_error = match e {
+        sov_api_spec::Error::ErrorResponse(response_value) => response_value,
+        _ => return false,
+    };
+
+    // Get the error string from the details map
+    let error_str = match api_error.details.get("error") {
+        Some(serde_json::Value::String(s)) => s,
+        _ => return false,
+    };
+
+    // Parse the error string to extract the actual rollup height
+    // Pattern: "Block state assertion failed at rollup height <actual_height>."
+    let actual_height = match extract_rollup_height_from_error(&error_str) {
+        Some(h) => h,
+        None => return false,
+    };
+
+    // If the actual height is greater than expected, this is a "too late" error
+    actual_height > expected_height
+}
+
+/// Extract the rollup height from the error message.
+/// Pattern: "Block state assertion failed at rollup height <height>."
+fn extract_rollup_height_from_error(error_str: &str) -> Option<u64> {
+    // Find "Block state assertion failed at rollup height "
+    let prefix = "Block state assertion failed at rollup height ";
+    let start_idx = error_str.find(prefix)?;
+    let height_start = start_idx + prefix.len();
+
+    // Find the next non-digit character (should be '.')
+    let remaining = &error_str[height_start..];
+    let height_str = remaining.split(|c: char| !c.is_ascii_digit()).next()?;
+
+    height_str.parse::<u64>().ok()
+}
+
+pub async fn state_validation_worker(
+    client: sov_api_spec::Client,
+    rx: watch::Receiver<bool>,
+) -> anyhow::Result<()> {
+    // Subscribe to slots
+    let mut slot_stream = client
+        .subscribe_slots_with_children(IncludeChildren::new(false))
+        .await?;
+
+    tracing::info!("State validation worker started");
+
+    while !*rx.borrow() {
+        let latest_slot = match slot_stream.next().await {
+            Some(Ok(slot)) => slot,
+            Some(Err(e)) => {
+                tracing::error!("Error receiving slot: {}", e);
+                continue;
+            }
+            None => break,
+        };
+        let latest_slot = drain_slot_stream(&mut slot_stream, latest_slot);
```

Comment:
Nit (not blocking): Why in line 246 we handle `Some(Err(e))` case explicitly but we skip this case inside `drain_slot_stream`?. Would it be possible to move the code from line 244 to 251 into `drain_slot_stream`
So, all the slot processing logic is contained in one method?